### DISCUSSION TITLE
execution-time in bash

### DISCUSSION THREAD

**Discussion by MatthewBrien** [JanDeDobbeleer/oh-my-posh#474] [Troubleshoot] [answered]
I'd like to have the execution time of my previous command in my prompt, but I'm not quite sure how to pass that parameter to oh-my-posh.

The Docs say "the installation guide shows how to include this argument for PowerShell and Zsh."  But I don't actually see an example for either of those shells.
I tried to reverse engineer the module, but my go is not very good. 

As a proof of concept, In my .bashrc file I've overridden PS0 to save the start time, and added a function to call when the prompt is printed, which I call in the prefix field of the executiontime block of my .oh-my-posh.omp.json concept.

Is there a specific environmental variable I can overwrite to give oh-my-posh the execution time?
Or is execution time only supported in PowerShell and Zsh? 

.bashrc
`TIMER_START="/tmp/${USER}.start.$$"

PS0='$(echo `date +%s%N` >  $TIMER_START)'

function get_execution_time() {
        if [[ -f $TIMER_START ]]; then
            local __END=`date +%s%N`
            local __START=$(cat "$TIMER_START")
            let __DURATION=$(((($__END-$__START))))
            echo $__DURATION
        fi

}
eval "$(oh-my-posh --init --shell bash --config ~/.oh-my-posh.omp.json )"
`

.oh-my-posh.omp.json
`        {
          "type": "executiontime",
          "style": "powerline",
          "powerline_symbol": "\uE0B0",
          "foreground": "#ffffff",
          "background": "#8800dd",
          "properties": {
            "threshold": 1,
            "style": "austin",
            "always_enabled": true,
            "prefix": " <#fefefe>\ufbab $(get_execution_time)ns</> "
          }`
          
Proof of Concept
![image](https://user-images.githubusercontent.com/6379561/109576390-a88cab00-7ac1-11eb-9250-ab1b67bcd2c8.png)

Thanks in advance!

**Comment by JanDeDobbeleer** [comment]
We tried to facilitate this, not sure why we didn't in the end (think we had trouble with it). If you have a way, extending the [init](https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/src/init/omp.bash) script is the way to go.

**Reply by JanDeDobbeleer** [reply]
@MatthewBrien in theory [this](https://github.com/JanDeDobbeleer/oh-my-posh/pull/477) should work, but even after updating bash it only works for the first prompt. Maybe I'm doing something silly:

```bash
export POSH_THEME=/Users/jan/.jandedobbeleer.omp.json

TIMER_START="/tmp/${USER}.start.$$"

function bash_setstarttime() {
  /usr/local/bin/oh-my-posh --millis > $1
}

PS0="$(bash_setstarttime $TIMER_START)"

function _update_ps1() {
    omp_elapsed=-1
    if [[ -f $TIMER_START ]]; then
        omp_now=$(/usr/local/bin/oh-my-posh --millis)
        omp_start_time=$(cat "$TIMER_START")
        omp_elapsed=$(($omp_now-$omp_start_time))
    fi
    PS1="$(/usr/local/bin/oh-my-posh --config $POSH_THEME --error $? --execution-time $omp_elapsed)"
}

if [ "$TERM" != "linux" ] && [ -x "$(command -v /usr/local/bin/oh-my-posh)" ]; then
    PROMPT_COMMAND="_update_ps1; $PROMPT_COMMAND"
fi

function runonexit() {
  rm $TIMER_START
}

trap runonexit EXIT
```

**Reply by MatthewBrien** [reply]
I think I cracked it!
I was having an issue where the previous command's execution time was repeated until I had a long running command.
I resolved this by removing the TIMER_START file after reading it.

```bash
export POSH_THEME=/home/matthewbrien/.oh-my-posh.omp.json
if [ "$TERM" != "linux" ] && [ -x "$(command -v /usr/local/bin/oh-my-posh)" ]; then
    PROMPT_COMMAND="_update_ps1; $PROMPT_COMMAND"
fi

TIMER_START="/tmp/${USER}.start.$$"

PS0='$(date +%s%N > $TIMER_START)'

function _update_ps1() {
__EXIT=$?
__DURATION=0
if [[ -f $TIMER_START ]]; then
local __END=$(date +%s%N)
local __START=$(cat "$TIMER_START")
 __DURATION=$(((($__END-$__START))/1000000))

rm $TIMER_START
fi
PS1="$(/usr/local/bin/oh-my-posh   --config $POSH_THEME   --error $__EXIT --execution-time $__DURATION )"
}

```

But the real reason I was stuck was because I had added  eval "$(oh-my-posh --init --shell bash --config ~/.oh-my-posh.omp.json )" to my .bashrc file so I didn't even know the_update_ps1 function was running, and  I was trying to add --exection-time to the oh-my-posh --init line. 😆

Thanks for your help!

**Reply by JanDeDobbeleer** [reply]
@MatthewBrien the [next](https://github.com/JanDeDobbeleer/oh-my-posh/pull/477) version of OMP will have this baked in. Thanks for making it better!

<img width="660" alt="image" src="https://user-images.githubusercontent.com/2492783/109769011-e987db00-7bf9-11eb-92cf-ad91e941bb44.png">

**Comment by JanDeDobbeleer** [comment, marked as answer]
Supported as of [v3.105.0](https://github.com/JanDeDobbeleer/oh-my-posh/releases/tag/v3.105.0)

**Reply by MatthewBrien** [reply]
![image](https://user-images.githubusercontent.com/6379561/109813681-ac244d00-7bfb-11eb-92b7-9c3d20689cef.png)
